File: algorithms/algorithms-cplusplus/leetcode/Question_0454_4Sum_II.cc
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
    int res = 0;
    unordered_map<int, int> m;
    for(auto i : A){
      for(auto j : B){
        m[i + j]++;
      }
    }        
    for(auto i : C){
      for(auto j : D){
        if(m.count(-(i + j))){
          res += m[-(i + j)];
        }
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0454_4Sum_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.HashMap;
import java.util.Map;

public class Question_0454_4Sum_II {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i : nums1) {
            for (int j : nums2) {
                if (map.containsKey(i + j)) {
                    map.put(i + j, map.get(i + j) + 1);
                } else {
                    map.put(i + j, 1);
                }
            }
        }
        int res = 0;
        for (int i : nums3) {
            for (int j : nums4) {
                if (map.containsKey(-(i + j))) {
                    res += map.get(-(i + j));
                }
            }
        }
        return res;
    }
}
